### PR TITLE
uplift probabilities refactoring 

### PR DESCRIPTION
**Pull request by RobertChin84**
This PR changes the probabilities to normalize within group rather than over the both groups.

`P(active|control) = N(active and control)/N(control) `

as opposed to 

`P(active|control) = N(active and control)/N(control + treated)`

Modified the following:
*  tree.py with a `predict_uplift` method
* forest.py with a `predict_uplift` method

You can reference this sklearn branch with 

`scikit-learn==0.17.1-RUBIKLOUD-2`

### CODE REVIEW

**Comment by apetresc** [comment]
Hey @RobertChin84, I've rebased your branch to hopefully contain the right set of commits. The diffs should now reflect your changes - let me know if they're the right ones :)

**Comment by RobertChin84** [comment]
Hi @apetresc  I have had a look at the `uplift_probability_corrections` branch and everything looks fine :)! Do you want to go ahead and do a merge?

File: sklearn/ensemble/forest.py [hunk 1]
```diff
@@ -597,6 +597,64 @@ def predict_log_proba(self, X):
...
+            The class probabilities of the input samples. The order of the
+            classes corresponds to that in the attribute `classes_`.
+        """
+        # Check data
+        X = self._validate_X_predict(X)
+
+        # Assign chunk of trees to jobs
+        n_jobs, _, _ = _partition_estimators(self.n_estimators, self.n_jobs)
+
+        # Parallel loop
```
Comment (bjlkeng) [inline sklearn/ensemble/forest.py:622]: Can you fix these comments?  No new line between comments and the block below.

Same with "# Reduce" below.

File: sklearn/tree/tree.py [hunk 2]
```diff
@@ -727,6 +727,73 @@ def predict_log_proba(self, X):
...
+        -------
+        p : array of shape = [n_samples, n_classes], or a list of n_outputs
+            such arrays if n_outputs > 1.
+            The uplift probabilities of the input samples. The order of the
+            classes corresponds to that in the attribute `classes_`.
+        """
+        X = self._validate_X_predict(X, check_input)
+        proba = self.tree_.predict(X)
+
+        if self.n_outputs_ == 1:
```
Comment (bjlkeng) [inline sklearn/tree/tree.py:770]: You should put an assert here to check that n_outputs == 4 I think.  It should always be 4 in the uplift scenario.

File: sklearn/ensemble/forest.py [hunk 3]
```diff
@@ -597,6 +597,64 @@ def predict_log_proba(self, X):
...
+        all_proba = Parallel(n_jobs=n_jobs, verbose=self.verbose,
+                             backend="threading")(
+            delayed(_parallel_helper)(e, 'predict_uplift', X,
+                                      check_input=False)
+            for e in self.estimators_)
+        # Reduce
+
+        proba = all_proba[0]
+
+        if self.n_outputs_ == 1:
```
Comment (bjlkeng) [inline sklearn/ensemble/forest.py:631]: Same comment about n_outputs_ == 4

**Review by bjlkeng — CHANGES_REQUESTED** [review]
Math looks correct, just some minor changes below.  Please make sure all the scikit tests pass (including the unit test I added for uplift).

File: sklearn/tree/tree.py (hunk 2 above)
Comment (RobertChin84) [reply sklearn/tree/tree.py:770]: as per our discussion - I will place an asset for the number of classes.

**Comment by RobertChin84** [comment]
Updated as per your suggestions @bjlkeng.

**Review by bjlkeng — APPROVED** [review]
